docs_src/design/adr/device-service/0012-DeviceService-Filters.md

```diff
@@ -0,0 +1,83 @@
+
+- based on the reading value (numeric) of a reading outside a specified range (min/max) described in the device profile for a device resource.  Thus avoiding sending in outlier or jittery data readings that could negatively effect analytics.
+- based on the reading value (numeric) equal to or near (with in some specified range) the last reading.  This allows a device service to reduce sending in event/readings that do not represent any significant change.  This filter would require that the device service maintains some sort of cache of past readings.
+
+If the concept of shared functions is achieved, there is no reason why these functions could not be reusable in application service pipelines as well.
+
+### Design Considerations
+
+#### Pipeline Nada
+
+As Device Services do not have the concept of a functions pipeline like application services do, consideration must be given as to how and where to:
+
+- invoke the filtering functions
+- provide configuration to specify which filter functions to invoke
+
+#### Function Inflection Point
+
+When instructed to "get" new readings, the function [execReadDeviceResource](https://github.com/edgexfoundry/device-sdk-go/blob/0bbcb663a9153978e7e9ef8c297d5988e58906d0/internal/handler/command.go) is called which subsequently calls on the the device service driver's HandleReadCommands function to get the latest sensor values from the device.  The driver's HandleReadCommands returns the sensor reading data (via array of CommandValues) to be put into an event.
+
+After receiving the CommandValues the execReadDeviceResource function calls the cvsToEvent function convert the CommandValues into and event/reading objects (from go-mod-core-contracts) and returns this model to be sent via REST to core data by the rest of the SDK.
+
+It is precisely after the convert to event/readings and before returning that result in execReadDeviceResource function that the device service should invoke the required filter functions.
+
+*Are the Go and C SDKs similar enough to use this design?*
+*Do we care about function order?*
+
+#### Function Configuration
+
+While device services do not have pipelines, the inclusion and configuration of filters for device services could take on a similar look (to provide symmetry with app services)
+
+``` toml
+[Writable.Functions.FilterByDeviceName]
+    [Writable.Functions.FilterByDeviceName.Parameters]
+    DeviceNames = "Random-Float-Device,Random-Integer-Device"
+    FilterOut = "false"
+[Writable.Functions.FilterByValueDescriptor]
+    [Writable.Functions.FilterByValueDescriptor.Parameters]
+    ValueDescriptors = "RandomValue_Int8, RandomValue_Int64"
+    FilterOut = "false"
+```

```

Comment:
Not sure is configuration is needed in this way.

Reply:
I suggest putting these into DeviceService object in core-metadata, and provide sub object APIs to control them.  We would like to make the filter configurable.

Reply:
Putting the idea out there so as to get reaction and find the best approach, so I encourage thoughts around the best way to do this (and where).  What I would like us to consider (if not implement in crawl phase) is how do we make it so that DS or App Service configuration of filters shares the same constructs so users see continuity in approach/setup no matter where they use the filters.  This is how app services does it today, so I borrowed that from app services.  
Someone want to provide an alternate config layout and where/how it is managed?